components/proxy/src/test/kotlin/com/hotels/styx/routing/handlers/HostProxyTest.kt

```diff
@@ -65,10 +73,51 @@ class HostProxyTest : FeatureSpec() {
                     client?.sendRequest(any())
                 }
             }
+
+            scenario("Updates metrics for cancelled responses") {
+                val client = mockk<StyxHostHttpClient>()
+                val originMetrics = mockk<OriginMetrics>()
+
+                every { client.sendRequest(any()) } returns Eventual(Mono.never())
+
+                val hostProxy = HostProxy(HostAndPort.fromString("abc:80"), client, originMetrics)
+
+                hostProxy.handle(get("/").build())
+                        .toMono()
+                        .subscribe()
+                        .dispose()
+
+                verify(exactly = 1) { originMetrics.requestCancelled() }
+            }
+
+            scenario("Updates metrics for cancelled response content") {
+                val client = mockk<StyxHostHttpClient>()
+                val originMetrics = mockk<OriginMetrics>()
+
+                every { client.sendRequest(any()) } returns Eventual.of(
+                        LiveHttpResponse
+                                .response()
+                                .body(ByteStream(Flux.never()))
+                                .build())
+
+                val hostProxy = HostProxy(HostAndPort.fromString("abc:80"), client, originMetrics)
+
+                hostProxy.handle(LiveHttpRequest.get("/").build(), HttpInterceptorContext.create())
+                        .toMono()
+                        .block()
+                        .let { response ->
+                            response.body()
```

Comment:
My IDE says `Unsafe use of a nullable receiver of type LiveHttpResponse?` Do you know what this means, and is it correct?

Reply:
It is because compiler cannot prove that `response` won't be null under any circumstances.

Being correct we should use Kotlin's `?.` operator, but this then propagates further down the chain:
                            response?.body()
                                    ?.toFlux()
                                    ?.subscribe()
                                    ?.dispose()

This being test, it doesn't if we throw an NPE so I'd leave the code as it is.

Reply:
Actually, I will use a null assertion operator: `!!.`:

                            response!!.body()
                                    .toFlux()
                                    .subscribe()
                                    .dispose()